text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,1243 @@
+a uniform distribution.
+
+[`AsciiWordChar`] is currently an oddity, used in many tests but with no hard
+requirements on form or function. This could be renamed and augmented in line
+with [Regex character classes](https://en.wikipedia.org/wiki/Regular_expression#Character_classes).
+
+#### Range
+
+There is one further uniform distribution:
+
+*   [`Range`] specifies uniform distribution over a range `[a, b)` and supports
+    integer and floating-point types
+
+This [`Range`] is minimally changed from the current `rand`, and supports
+extension to user-defined types by exposing its internal fields.
+
+An alternative
+implementation, [`range2`], has been written in an attempt to improve extension
+to other types and avoid the need for an unused `zone` field with float types,
+but has some drawbacks, perhaps most notably that `Range` is parameterised so
+that `Range::new(low, high)` must be replaced with `range(low, high)` or
+`Range::<T>::new(low, high)`.
+
+**Question:** which `range` implementation should we choose?
+
+Unfortunately `range2` exposes more public types like `RangeInt<X>` and
+`RangeFloat<X>`, and must retain the `SampleRange` trait but *only* to identify
+types for which a `Range` implementation is available. (Suggestions to improve
+this code welcome.)
+
+Note that while `Range` is open to implementation for user-defined types, its
+API with a "low" and "high" may not be appropriate for many types; e.g. a
+complex type may want "low_real", "high_real", "low_complex" and "high_complex"
+parameters. For such uses, it is suggested the user create a new distribution
+(e.g. `ComplexRange`) and not try to extend `Range`.
+
+#### Non-uniform distributions
+
+Finally, there are several [`distributions`]
+unchanged from the current `rand`:
+
+*   `Exp`
+*   `Normal`, `LogNormal`
+*   `Gamma`, `ChiSquared`, `FisherF`, `StudentT`
+
+Currently these are only implemented for `f64`. They could be extended to `f32`
+but this might require adding some more lookup tables to the crate.
+
+Internally, `Exp(1)` and `N(0, 1)` (standard normal) fixed distributions are
+used; these could be exposed via new zero-size distribution structs.
+This might be slightly faster for some uses (avoid a multiplication and extra
+data access).
+
+Most distributions are implemented in public sub-modules, then *also* imported
+into `distributions` via `pub use`. Possibly the sub-modules should be hidden.
+
+#### Conversion to floating point
+
+Currently this is implemented via `impl Distribution<f32> for Uniform01` and
+the `f64` equivalent in the strawman revision, and within the `Rng` trait in
+the current `rand`. It has been suggested that this should be implemented in
+a simple function (used by `Rand` / `Uniform01`) so that users only wanting to
+use a small subset of the library for cryptography do not need to use the
+distributions code. This is only really useful if the `rand` crate is split
```

Comment:
Should Rand expose conversion functions or not?
You could see them as building blocks. I think they can be useful.
You may want to use them on an already generated random number.
Or use them to implement your own distribution or range code. I don't expect Rand to include all possible distributions. Or to have the most performant solution for every scenario.
For example, I couldn't get the range code to work with higher precision using just the distributions without regressing the benchmarks.

Maybe something like this can be exposed:
```
pub trait FloatConversions {
    type F;

    /// Uniformly distributed half-open range [0,1).
    /// The precision gets higher closer to zero, up to 2^-32 for f32 and 2^-64 for f64.
    fn closed_open01(self) -> Self::F;

    /// Uniformly distributed half-open range (0,1].
    /// The precision gets higher closer to zero, up to 2^-32 for f32 and 2^-64 for f64.
    fn open_closed01(self) -> Self::F;

    /// Uniformly distributed half-open range [-1,1).
    /// The precision gets higher closer to zero, up to 2^-32 for f32 and 2^-64 for f64.
    fn closed_open11(self) -> Self::F;

    /// Uniformly distributed half-open range [0,1).
    /// The precision is fixed, with 2^-23 for f32 and 2^-52 for f64.
    /// This leaves a couple of the random bits available for other purposes
    /// (the 9 rightmost bits for f32, and 12 for f64).
    fn closed_open01_fixed(self) -> Self::F;

    /// Uniformly distributed half-open range (0,1].
    /// The precision is fixed, with 2^-23 for f32 and 2^-52 for f64.
    /// This leaves a couple of the random bits available for other purposes
    /// (the 9 rightmost bits for f32, and 12 for f64).
    fn open_closed01_fixed(self) -> Self::F;

    /// Uniformly distributed half-open range [-1,1).
    /// The precision is fixed, with 2^-22 for f32 and 2^-51 for f64.
    /// This leaves a couple of the random bits available for other purposes
    /// (the 9 rightmost bits for f32, and 12 for f64).
    fn closed_open11_fixed(self) -> Self::F;
}
```

Reply:
I want to say yes, but where precisely? In a new module alongside the `Uniform01` distribution maybe? (Other readers: [this has already been discussed a little](https://github.com/dhardy/rand/pull/1).)

Possibly the comments should refer to the "9 least significant bits" instead of *rightmost*.

Reply:
"9 least significant bits" sounds much better :smiley:. How about a new top level model `utils` or `misc`? Not that I think it is that important. More the opposite. But I don't really like deep hierarchies. And I can imagine there may be more functions that don't really fit anywhere else.

That leaves `distribution` for functions that can choose for themselves when and how many random numbers to generate to get their distribution. If I look at the number of distributions in a similar [module in Julia](http://juliastats.github.io/Distributions.jl/univariate.html) there may be some room to grow in the future...